runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java

```diff
@@ -0,0 +1,382 @@
+          throws Exception {
+    PTransform transformToUpgrade =
+        runnerAPIpipeline.getComponents().getTransformsMap().get(transformId);
+    if (transformToUpgrade == null) {
+      throw new Exception("Could not find a transform with the ID " + transformId);
+    }
+
+    ByteString configRowBytes =
+        transformToUpgrade.getAnnotationsOrThrow(PTransformTranslation.CONFIG_ROW_KEY);
+    ByteString configRowSchemaBytes =
+        transformToUpgrade.getAnnotationsOrThrow(PTransformTranslation.CONFIG_ROW_SCHEMA_KEY);
+    SchemaApi.Schema configRowSchemaProto;
+    try {
+      configRowSchemaProto =
+          (SchemaApi.Schema)
+              new ObjectInputStream(new ByteArrayInputStream(configRowSchemaBytes.toByteArray()))
+                  .readObject();
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+
+    Row configRow =
+        RowCoder.of(SchemaTranslation.schemaFromProto(configRowSchemaProto))
+            .decode(new ByteArrayInputStream(configRowBytes.toByteArray()));
+    ByteStringOutputStream outputStream = new ByteStringOutputStream();
+    try {
+      RowCoder.of(configRow.getSchema()).encode(configRow, outputStream);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    ExternalTransforms.ExternalConfigurationPayload payload =
+        ExternalTransforms.ExternalConfigurationPayload.newBuilder()
+            .setSchema(configRowSchemaProto)
+            .setPayload(outputStream.toByteString())
+            .build();
+
+    RunnerApi.PTransform.Builder ptransformBuilder =
+        RunnerApi.PTransform.newBuilder()
+            .setUniqueName(transformToUpgrade.getUniqueName() + "_external")
+            .setSpec(
+                RunnerApi.FunctionSpec.newBuilder()
+                    .setUrn(transformToUpgrade.getSpec().getUrn())
+                    .setPayload(ByteString.copyFrom(payload.toByteArray()))
+                    .build());
+
+    for (Map.Entry<String, String> entry : transformToUpgrade.getInputsMap().entrySet()) {
+      ptransformBuilder.putInputs(entry.getKey(), entry.getValue());
+    }
+    for (Map.Entry<String, String> entry : transformToUpgrade.getOutputsMap().entrySet()) {
+      ptransformBuilder.putOutputs(entry.getKey(), entry.getValue());
+    }
+
+    ExpansionApi.ExpansionRequest.Builder requestBuilder =
+        ExpansionApi.ExpansionRequest.newBuilder();
+    ExpansionApi.ExpansionRequest request =
+        requestBuilder
+            .setComponents(runnerAPIpipeline.getComponents())
+            .setTransform(ptransformBuilder.build())
+            .setNamespace(UPGRADE_NAMESPACE)
+            .build();
+
+    ExpansionApi.ExpansionResponse response =
+        clientFactory.getExpansionServiceClient(transformServiceEndpoint).expand(request);
+
+    if (!Strings.isNullOrEmpty(response.getError())) {
+      throw new IOException(String.format("expansion service error: %s", response.getError()));
```

Comment:
IOException doesn't seem to be the right thing to throw for a mis-configured transform.